### ISSUE TITLE
`rustc_private`?

### ISSUE THREAD

**Issue by smoelius** [rust-lang/docs.rs#2088]
Can a crate that uses `rustc_private` be built on docs.rs?

I have been trying to build this crate: https://docs.rs/crate/dylint_linting

My most recent attempt was to [add the relevant `rustc` components manually](https://github.com/trailofbits/dylint/blob/ba7be1fc883e1a36ebf52827cf6c9bb7cdefb8e2/utils/linting/build.rs#L40-L44) from the build script, which [didn't work](https://docs.rs/crate/dylint_linting/2.1.7/builds/775903).

Is what I am trying to do possible? If so, what is the "right" way to build a crate that uses `rustc_private` on docs.rs?

Thank you.

**Comment by jyn514** [comment]
Hmm. This is definitely not supported today, but it seems ok to run `rustup component add rustc_private` before building the docs. My only concern is that it seems like most builds will fail *anyway* because we always build with latest nightly and the API changes so frequently.

@smoelius can you talk a little more about how you were imagining this would work? "only the latest nightly" seems limiting; I guess we could support an arbitrary date but then it will be easy to overload the server just by publishing many crates with different rustc_private toolchains.

cc #506, https://github.com/rust-lang/docs.rs/issues/228

**Comment by smoelius1** [comment]
> @smoelius can you talk a little more about how you were imagining this would work? "only the latest nightly" seems limiting

"Only the latest nightly" would work for my use case.

Strictly speaking, `dylint_linting` has always required a nightly compiler. The intent is that it build with every nightly since the one coinciding with it's initial release. [Here](https://github.com/trailofbits/dylint/blob/a1897cb988167bc7cf136206d87cc7036d6097ab/utils/linting/src/lib.rs#L228), for example, is a case where `dylint_linting` had to adjust for a rustc API change.

`dylint_linting` has been building successfully on docs.rs because it consists largely of macros. It only started requiring `rustc_private` recently.

If the latest nightly was not possible, an arbitrary recent nightly chosen by docs.rs also seems workable.

In other words, if docs rs' policy was "we'll give you a nightly with `rustc_private` from within (say) the past six weeks, the rest is up to you," I could live with that.

(Apologies as I am traveling and don't have access to my normal @smoelius account. If you would prefer, I could reaffirm the above this coming Monday.)

**Comment by jyn514** [comment]
> "Only the latest nightly" would work for my use case

Oh, perfect! that should be quite easy to do on our side then :)

> Apologies as I am traveling and don't have access to my normal @smoelius account. If you would prefer, I could reaffirm the above this coming Monday.

No worries, you're perfectly fine :) I'm not worried about someone maliciously pretending to be you to tell us to implement a feature lol (not sure how that would even work).

~~~

Mentoring instructions: add a call to `toolchain.install_component("rustc_private")` around here, following rustfmt's example:
https://github.com/rust-lang/docs.rs/blob/e3e044be64c8a26b98e9677d02f5a2b31367a6a4/src/docbuilder/rustwide_builder.rs#L191
